### PR TITLE
manifest: Add bootupd for x86_64|aarch64

### PR DESCRIPTION
**Pull request by cgwalters**
This converts the EFI update directory from (rpm-)ostree specific
layout to bootupd layout.  Required for both initial
provisioning with bootupd as well as later updates.

Required by: https://github.com/coreos/coreos-assembler/pull/1695

### CODE REVIEW

**Comment by cgwalters** [comment]
OK I reworked this so that it's a no-op if bootupd isn't installed.  Hopefully we can merge this now and testing out bootupd will involve less juggling of git branches.

**Comment by jlebon** [comment]
I'm having trouble understanding how this all ties together. I would've expected bootupd on the compose side to interact in coreos-assembler somewhere in `create_disk.sh`. Doesn't it need to fiddle with `/boot`?

**Comment by cgwalters** [comment]
> I'm having trouble understanding how this all ties together.

No worries, me too :smile:   I messed this all up initially.

Basically this part replaces `/usr/lib/ostree-boot`.

> I would've expected bootupd on the compose side to interact in coreos-assembler somewhere in create_disk.sh. Doesn't it need to fiddle with /boot?

That's a separate phase in https://github.com/coreos/coreos-assembler/pull/1695

*This* phase needs to happen for every single ostree commit build.  The coreos-assembler one happens only when we make a bootimage.

**Comment by cgwalters** [comment]
Side note: the reason this is useful now is I am testing bootupd via
`make install DESTDIR=/var/srv/walters/builds/fcos/overrides/rootfs/` so it doesn't matter that it's not listed in the manifest too.  But I do plan to package it soon.

**Comment by cgwalters** [comment]
OK bootupd is in the repos; I reworked this to make it explicitly required for `x86_64` and `aarch64` (though I need to fix the RPM `ExclusiveArchitecture` for the latter still).

File: manifests/bootupd.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Integration with https://github.com/coreos/bootupd
+# xref https://github.com/coreos/fedora-coreos-tracker/issues/510
+packages:
+  - bootupd
```
Comment (jlebon) [inline manifests/bootupd.yaml:4]: Need a matching entry in the lockfiles for this.

File: manifests/bootupd.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+postprocess:
+  - |
+    #!/bin/bash
+    set -xeuo pipefail
+    /usr/bin/bootupctl backend generate-update-metadata /
```
Comment (jlebon) [inline manifests/bootupd.yaml:11]: Can you add a comment here with a short explanation of what this does? E.g. (IIUC):

```
# Transforms /usr/lib/ostree-boot into a bootupd-compatible update payload
```

**Review by jlebon — APPROVED** [review]
LGTM overall!

**Comment by cgwalters** [comment]
https://github.com/coreos/coreos-assembler/pull/1695#issuecomment-696241376
so that PR needs to land first.

**Comment by cgwalters** [comment]
OK, :heavy_check_mark: CI here.  We should probably debate whether to run with this right now; bootupd is still a WIP but I know it works for the "install" case, and having this in will help a lot developing it.  OTOH it does mean shipping it to all FCOS users in WIP status.

Hmm...maybe what we want is support for "build options" so that we can merge to testing-devel but turn off the option for stable?  Not sure.

**Comment by jlebon** [comment]
> We should probably debate whether to run with this right now; bootupd is still a WIP but I know it works for the "install" case, and having this in will help a lot developing it. OTOH it does mean shipping it to all FCOS users in WIP status.

I'm personally fine with merging it. We don't publicize/document the new CLI tool anywhere super user visible right now, and otherwise AFAICT bootupd on its own doesn't change the behaviour of a booted FCOS in any meaningful way, right?

Do you see any risks/potential areas for regression from this change?

File: manifests/bootupd.yaml (hunk 2 above)
Comment (jlebon) [reply manifests/bootupd.yaml:11]: Did you mean to address this one?

**Comment by darkmuggle** [comment]
> OTOH it does mean shipping it to all FCOS users in WIP status.

How much of a WIP is bootupd right now? Is it "alpha" or "beta" quality? If is the former, then I think we should hold off. And what's the risk of bricking a user? FCOS is stable, but we would be introducing something that unstable by definition into the code. 

> AFAICT bootupd on its own doesn't change the behaviour of a booted FCOS in any meaningful way, right

This PR enables the socket. I would call that a meaningful change. 

If we do no foresee a major refactor or changes, then I would be okay shipping this once we have Kola tests to make sure we're good. I think we're probably safe, but let's put up the guard rails first to catch our human mistakes.

**Comment by cgwalters** [comment]
> How much of a WIP is bootupd right now? Is it "alpha" or "beta" quality? If is the former, then I think we should hold off. 

I'd say alpha.

> And what's the risk of bricking a user?

If they somehow discover and run `bootupctl update` manually then non-zero, otherwise zero.  So...effectively zero.

> This PR enables the socket. I would call that a meaningful change.

That's mostly an implementation detail I think, it's a unix domain socket, not listening on TCP or anything like that.

But perhaps the simplest alternative is to keep the socket disabled?  That would make it easy to turn on via Ignition or by hand and test out.

File: manifests/bootupd.yaml (hunk 2 above)
Comment (cgwalters) [reply manifests/bootupd.yaml:11]: I put the comment above, but I guess it is better inline in the script.

**Comment by cgwalters** [comment]
> But perhaps the simplest alternative is to keep the socket disabled? That would make it easy to turn on via Ignition or by hand and test out.

Updated with that change.

File: overlay.d/15fcos/usr/lib/systemd/system-preset/45-fcos.preset [hunk 3]
```diff
@@ -3,3 +3,4 @@ enable fedora-coreos-pinger.service
 # Provide information if no ignition is provided
 enable coreos-check-ignition-config.service
 enable coreos-check-ssh-keys.service
+
```
Comment (jlebon) [inline overlay.d/15fcos/usr/lib/systemd/system-preset/45-fcos.preset]: Minor/optional: spurious new line.

**Review by jlebon — APPROVED** [review]
LGTM. Though to be safe let's hold until the current round of releases for this week is done. Otherwise, this'll get immediately promoted to `testing` and `next` e.g. tomorrow.

**Comment by cgwalters** [comment]
Also did https://github.com/coreos/bootupd/pull/27

**Comment by jlebon** [comment]
Releases are out now. Restarted CI here.

**Pull request merged**